### ISSUE TITLE
Unable to save link to node with non-ascii characters in title when transliteration is off

### ISSUE THREAD

**Issue by indigoxela** [backdrop/backdrop-issues#4630]
**Description of the bug**

The filter-format-editor-link-form ("Insert link" dialog) verifies urls before inserting it into the Content. It however doesn't handle urls with non-ascii characters.

**Steps To Reproduce**

1. Make sure transliteration of aliases is off on admin/config/urls/path/patterns/settings
2. Create a post or page with following title "Ärger im Büro"
3. Open another page/post, open the "Insert link" dialog and start typing "Ärger..."
4. The right node title and path alias shows up, but you can not insert the link

**Actual behavior**

The form validation refuses to insert that link.

![invalid-url-when-diacritics](https://user-images.githubusercontent.com/842066/93739019-8d9b3e00-fbe7-11ea-987a-217963f42954.png)

**Expected behavior**

This *is* a valid url: `http://somedomain.tld/posts/%C3%A4rger-im-b%C3%BCro`

It shows up in the browsers' address field like `http://somedomain.tld/posts/ärger-im-büro`

So Backdrop handles these urls correctly, but the dialog form validation fails for some reason.

**Additional information**

- Backdrop CMS version: 1.17.0
- Browser(s) and their versions: Firefox ESR or Chromium

After a rewrite of function _filter_format_editor_link_url_validate() the PR should also cover:

- https://github.com/backdrop/backdrop-issues/issues/4855
- https://github.com/backdrop/backdrop-issues/issues/4647
- https://github.com/backdrop/backdrop-issues/issues/4919

**Comment by olafgrabienski** [comment]
Confirmed! I got the same result when I followed the Steps To Reproduce.

**Comment by olafgrabienski** [comment]
Workaround for the time being:

- In the dialog, replace the umlauts from link destination with alternatives, e.g. `posts/arger-im-buro`, and insert the link.
- Switch to Source mode of CKEditor, and add the umlauts back: `<a href="/posts/ärger-im-büro">`.

**Comment by indigoxela** [comment]
The trick is to encode the paths. A [PR](https://github.com/backdrop/backdrop/pull/3331) is available for testing and review.

To test it, follow the steps in the issue description. Don't be astounded that a path like `/ärger-im-büro` turns into `/%C3%A4rger-im-b%C3%BCro` - that's necessary for a valid url.

If you need "a second opinion" take a look on how the German Wikipedia does it. Example: https://de.wikipedia.org/wiki/H%C3%A4moglobin

And also don't be astounded that browsers show it differently in their address bar.

And another thing: yes, browsers would "translate" an url like /ärger-im-büro to /%C3%A4rger-im-b%C3%BCro, but umlauts (and other utf8 stuff) aren't valid in urls. Correct me if I'm wrong, though.

**Comment by olafgrabienski** [comment]
I've finally had a look at the PR. The fix works for me: I was able to insert the link in step 4 of the STR.

There is one issue with the PR: If a linked page contains an umlaut, the leading slash of the link path is missing:

`<a href="%C3%A4rger-im-b%C3%BCro">`
`<a href="m%C3%B6glichkeiten">`

The missing slash can lead to Not found errors.

Compare a link to a page without umlaut which starts with a slash:
`<a href="/about">`

**Comment by indigoxela** [comment]
> There is one issue with the PR: If a linked page contains an umlaut, the leading slash of the link path is missing

There's a [related issue](https://github.com/backdrop/backdrop-issues/issues/4647), but I might be able to fix it for this use-case.

@olafgrabienski many thanks for testing.

**Comment by indigoxela** [comment]
@olafgrabienski happy testing again, my PR has been updated.

**Comment by olafgrabienski** [comment]
Thanks for the update, @indigoxela. I've tried these links, and all work for me in the PR's sandbox website:

```
<a href="/%C3%A4rger-im-b%C3%BCro">Ärger im Büro</a>
<a href="/about">About</a> page
<a href="/jede-menge-m%C3%B6glichkeiten">page which doesn't start with but contains an umlaut</a>
```

**Comment by indigoxela** [comment]
Based on the latest review comment, and as we recently discovered several (more or less) related issues with function _filter_format_editor_link_url_validate() - back to "needs work".

Possibly we'll expand the scope of this issue/PR to cover more of these problems. I'll have a closer look ASAP.

**Comment by indigoxela** [comment]
After having a much closer look, I realized that my former solution created a possible regression with urls containing query or fragment parts.

I'm usually a fan of baby-steps, but function _filter_format_editor_link_url_validate() causes too much trouble. So I ended up with a complete rewrite.

Benefit of it - this might also fix following related issues:

- https://github.com/backdrop/backdrop-issues/issues/4855
- https://github.com/backdrop/backdrop-issues/issues/4647
- https://github.com/backdrop/backdrop-issues/issues/4919

However, the bigger change needs thorough testing and possibly additional functional tests.

**Comment by indigoxela** [comment]
Functional tests have been added. They try to verify that:

- Non-ascii characters get encoded
- Language prefixes get added
- The `<front>` placeholder gets replaced
- Valid urls don't get mangled or rejected

But manual testing, that the problems discovered in this and the three related issues got fixed, is highly appreciated. Feedback is welcome.

**Comment by herbdool** [comment]
Thanks for all your work on this @indigoxela.

I am taking just an initial look at this. My feeling is that there seems to be too much of the logic stuffed into this function, where it seems to be trying to replicate and improve on `link_cleanup_url`.

I'm wondering if we should move `link_cleanup_url` out of the link module and put it into includes so it can be used by any module. Then you could improve that function instead. (And may need to move `link_validate_url` too.)

Though this is just an initial thought so I might be off base here.

**Comment by indigoxela** [comment]
@herbdool many thanks for having a look!

> I'm wondering if we should move link_cleanup_url out of the link module and put it into includes so it can be used by any module.

That's something that came to my mind, too. But then - this goes beyond the scope of this issue. I avoided to re-organize too much in core to have a chance to get the annoying editor link dialog bugs fixed. On the other hand... maybe the usage of that new validation function in the link module could be a follow-up?

@docwilmot many thanks for your helpful comments on the PR!

**Comment by indigoxela** [comment]
My PR has been updated with the changes requested by @docwilmot 

@herbdool I gave the idea to move link_cleanup_url / link_validate_url / _link_escape_domain out of the link module another try, but...
These three functions wouldn't solve the problem for the editor dialog. See #4919 for one of the reasons.

The main problem in the editor link dialog in its current form is, that it doesn't convert internal paths to functioning urls - including a possible language prefix, and that it doesn't handle non-ascii characters correctly (url() does all that). The essential part is the usage of form_set_value(). The autocomplete widget provides internal paths, the output must be a usable url. This is different with link fields - which use l() for the rendering (see the module's theme functions).

The functions used for validation in the overhauled _filter_format_editor_link_url_validate() are generic ones anyway (valid_url / backdrop_valid_path). So at least I didn't re-invent the wheel for validation. :wink:

**Comment by herbdool** [comment]
Thanks for the explanation @indigoxela. I've tested it in tugboat and it seems to work well. My feeling though is that the code still needs a bit of cleaning up to be easier to follow. I was playing around with the PR to figure out what's going on and how it could be easier to understand, but I paused to provide some feedback:

* the nested conditionals are hard to follow and I believe can be simplified by separating out the functionality that is cleaning up  the url from the functionality that is validating it. Perhaps a new `_filter_cleanup_url()`. 
* Adopting some "guard clauses" and avoiding the use of `if / else` as much as possible would help make it easier to follow. (See https://refactoring.guru/replace-nested-conditional-with-guard-clauses)
* Some functions in core still seem to have solved some things that could be used here, such as `_link_parse_url()` instead of `parse_url()` so it works with relative urls; `_link_parse_str()` instead of `parse_str()`. I don't know if that means those functions should be moved out or if we create versions of them specifically for filter module.

**Comment by indigoxela** [comment]
@herbdool many thanks for sharing your thoughts.

> the nested conditionals are hard to follow and I believe can be simplified by separating out the functionality that is cleaning up the url from the functionality that is validating it.

I wouldn't know how to further simplify the code. If you have an idea, please provide an alternate PR.

> Some functions in core still seem to have solved some things that could be used here

I already used all the functions from core that made sense to me. The functions from the link module didn't IMO.

Feel free to take over, but keep in mind, that we're trying to solve bugs, which currently prevent the link dialog from doing its job properly, resulting in wrong urls (404) inserted into content.

**Comment by docwilmot** [comment]
I understand @herbdool' s concern, but I suspect this one will be harder to "guard clause". We wouldn't want to use the `link_WHATEVER()` functions, since we're trying to remove a dependency on link module, as part of this issue. And IMO the code flow is actually not that complex, and I suspect more verbose and better-placed code comments would do the trick. Here's a first try which I think clears it up a bit:
```

/**
 * Element validation function.
 *
 * Additionally to validation, this function also rewrites values, if necessary.
 */
function _filter_format_editor_link_url_validate(&$element, &$form_state) {
  $value = trim($element['#value']);

  // Allow empty values, since this will not result in a link tag anyway.
  if ($value === '') {
    form_set_value($element, $value, $form_state);
    return;
  }

  $valid_url = FALSE;
  $parts = parse_url($value);

  // External urls have a host and/or a scheme. If the scheme is missing from
  // the URL, we check the syntax works if we prepend a common scheme.
  if (isset($parts['host'])) {
    if (valid_url($value, TRUE)) {
      $valid_url = TRUE;
    }
    //
    elseif (!isset($parts['scheme'])) {
      if (valid_url('http:' . $value, TRUE)) {
        $valid_url = TRUE;
      }
    }
  }
  // Otherwise this is an internal URL.
  else {
    $path = isset($parts['path']) ? $parts['path'] : '';
    // Remove the path from the array, keep possible query or fragment for
    // later usage as options in url().
    unset($parts['path']);
    if (isset($parts['query'])) {
      parse_str($parts['query'], $query_array);
      $parts['query'] = $query_array;
    }

    // We check the URL to see if it an existing path. If not, perhaps it
    // is missing the base path or needs transliteration.
    if (backdrop_valid_path($value)) {
      // This is a valid internal path like "node/1" or "<front>" and it's
      // accessible. Let url() add the base path (and language prefix).
      $value = url($path, $parts);
      $valid_url = TRUE;
    }
    else {
      // The syntax is valid for a relative url, but we still might have to
      // add the base path.
      if (valid_url($value, FALSE)) {
        if (substr($value, 0, 1) != '/' && substr($value, 0, 1) != '#') {
          $value = url($path, $parts);
        }
        $valid_url = TRUE;
      }
      // Finally if we've gotten this far, maybe the URL needs to be
      // transliterated or is actually invalid. Let url() encode the path
      // part, then we verify the reassembled url again.
      elseif (backdrop_encode_path($path) != $path) {
        $encoded = url($path, $parts);
        if (valid_url($encoded, FALSE)) {
          $value = $encoded;
          $valid_url = TRUE;
        }
      }
    }
  }

  if (!$valid_url) {
    form_error($element, t('The URL %url is not valid.', array('%url' => $value)));
  }

  form_set_value($element, $value, $form_state);
}
```

**Comment by herbdool** [comment]
I hope to get back to this soon. @docwilmot you only responded to a couple of my suggestions so I'll just mention that I also suggested moving the logic into a new function or two and calling them from this one. And if relying on link_* (if that makes sense) to actually move them out of the link module. Anyway, I was part way through trying that out.

**Comment by indigoxela** [comment]
> I hope to get back to this soon.

I'm looking forward to see your alternate approach. Especially how you reconcile the different demands in the Link and Filter module.
(It's not that I didn't try to combine things for those two modules, but I realized that the use-cases are too different to actually benefit from commonly used functions.)

**Comment by indigoxela** [comment]
> I hope to get back to this soon.

@herbdool when is "soon"? We're currently stuck, which is a bit frustrating when there's already a functioning PR around. I know, I'm impatient, but the bugs addressed here are really annoying and I'd like to see them fixed ASAP.

**Comment by herbdool** [comment]
I hope to give it a try this week. Hasn't been the best month for doing extra curricular activities.

**Comment by herbdool** [comment]
I'm working on it. Hopefully have something to show by weekend.

**Comment by herbdool** [comment]
@indigoxela here's my variation: https://github.com/backdrop/backdrop/pull/3552. I've added notes on the PR. I've obviously borrowed heavily from yours and adapted it.

**Comment by indigoxela** [comment]
@herbdool many thanks for working on this, I closed my PR to not confuse testers. ... And left a (sort of) question on your PR regarding encoding. Hopefully I'll find the time to do more detailed testing soon.

**Comment by indigoxela** [comment]
I didn't do thorough testing yet, but found some thingies, that either need clarification or work - I assume the latter.

**Comment by herbdool** [comment]
The recent releases for d7 Link have me thinking if we should be consistent with it. It's trying to save links as their internal path and not as aliases with possible prefixes. https://www.drupal.org/project/link/issues/836710

**Comment by indigoxela** [comment]
> The recent releases for d7 Link have me thinking if we should be consistent with it. It's trying to save links as their internal path and not as aliases with possible prefixes

I suspect, this won't work. The link module does conversion when rendering, the editor dialog inserts the link url as provided. So there's a fundamental difference in concepts. We have no output filter for that when rendering the content, so the url must be functioning and complete when inserted into the editor.

**Comment by herbdool** [comment]
That crossed my mind too. I was wondering if it might help in some cases, such as with the localization paths, to leave it up to the filter rendering to create an aliased path. Still just considering it. We could leave it for a follow-up issue (which would mean we take out the localization function from this one).

The downsides of making an aliased path on input is that it is less transferable than using the system path. node/1 will work even if changing the website from using language prefixes to domain switching. Also, it's hard to make it optional in case someone wants to turn it off.

It might be unusual to have a new filter setting but it could work. Plus it's optional in case someone wants to stick with node/1 instead of render it as de/something.

**Comment by indigoxela** [comment]
@herbdool I re-opened my [PR](https://github.com/backdrop/backdrop/pull/3331)

Let's put our heads together one more time to get out the best of both approaches.

**Comment by herbdool** [comment]
After refreshing my memory of the current functionality which saves the alias in most cases (except for nodes from other locale), I've created a different issue for making it easier to handle aliases: https://github.com/backdrop/backdrop-issues/issues/5004. So my last couple comments can be ignored on this issue - we'll see what happens to the other one.

And I've updated my PR so that if an external domain does not have a protocol, nor a relative protocol, then it'll add a protocol using http or https depending on the current site's setting.

**Comment by indigoxela** [comment]
@herbdool many thanks for the update.

I'll try to test/review this soon, but I'd like to raise one concern: It seems to me, your PR has derailed a bit. IMO it tries to solve too many different things at once. Complicated things. Maybe that's the reason, why I find it harder and harder to read and understand.

**Comment by herbdool** [comment]
Maybe you can explain more. The only thing I can think of that is not in your PR is trying to get the locale path. Everything else has the same result. The biggest difference is I split out the rewriting the URL from validation of the URL.

**Comment by indigoxela** [comment]
> Maybe you can explain more. ... is trying to get the locale path

Exactly that, function _filter_lookup_locale_path().

Splitting into smaller chunks / functions - that's understood. But your PR increases overall complexity.

Some comments definitely need an update, as they are misleading - given, I understand your code (I'm not so sure anymore :wink: ).

**Comment by indigoxela** [comment]
Some testing with multilingual:

If I use the autocomplete widget, everything works as expected. Unfortunately, if I use a (valid) path like `/de/node/1` this ends up as `/en/de/node/1`.

That's what I meant with "tries to solve too many different things". I am aware, how complex this harmless looking little "insert link" dialog in reality is. That's why I was skeptical. Sorry for my short message this morning, I should have verified my suspicion, before "nagging".

**Comment by herbdool** [comment]
I've removed `_filter_lookup_locale_path()`. It can be dealt with in a different issue.

**Comment by indigoxela** [comment]
> I've removed _filter_lookup_locale_path()

Yeah, urls like `/de/node/1` work properly again. Hm, we should probably add that url to the tests.

**Comment by indigoxela** [comment]
Meh. Just tested multilingual again out of curiosity and it failed...

- Enable language, locale, translation
- Set the "Detection and selection" to language prefix, and set a prefix for both languages "de", "en"
- Add a new post _with interface language English_: /en/node/add/post
- Add a link via dialog: /de/node/1 - which is valid
- Inspected in Editor source mode
- URL got inserted as "/en/de/node/1"

The code regarding that didn't change since my last attempt, so I probably tested wrong. I already suggested to add that check also to the automated tests in a PR comment.

**Comment by herbdool** [comment]
In your example is `/de/node/1` a node that is set to German? I assume so. I'll try replicate this.

**Comment by herbdool** [comment]
Also, when you're adding the new post which language is the interface in?

**Comment by herbdool** [comment]
Okay, trying your proposal. If it begins with a slash then it skips url(). This will also skip `//domain.tld` but that's fine.

**Comment by indigoxela** [comment]
> Okay, trying your proposal. If it begins with a slash then it skips url().

Great! It works! This time I double checked, that both languages were configured to use a prefix.

**Comment by indigoxela** [comment]
Who would have thought, that a tiny little dialog could turn out to be such a complex beast. :laughing: Many thanks to @herbdool for his perseverance.

Anyway, we're almost done now! The code is looking good.

People are encouraged to test again.

**Comment by indigoxela** [comment]
Friendly reminder @olafgrabienski (vacation, I know), @BWPanda and anyone else interested in a functioning link dialog: It took a lot of time and effort to overhaul that code, it's now back to "needs testing" - and stuck in this state for a while, so please help to revive it. It's such an annoying bug.

**Comment by olafgrabienski** [comment]
@indigoxela I've finally tested the PR using the sandbox website. Nearly all links work fine for me. I've tested the STR (all OK) and then the language related links discussed later:

- Node 1 link with language prefix and with starting slash ( /de/node/1 )
- Node 1 link with language prefix but without starting slash ( de/node/1 ) => see below
- Node 1 link without language prefix but with starting slash ( /node/1)
- Node 1 link without language prefix and without starting slash ( node/1 )
- Email link ( `mailto:boo@example.com` )
- External domain link without protocol ( `www.example.com` )
- External domain link with protocol ( `http://www.example.com` )

The current test pages (as long they're available):

- [Test page I](https://pr3552-gomcqa4wbiuielvzdtjmhpfnsc7jzfro.tugboat.qa/test-page-i), created while site was monolingual
- [Test page II](https://pr3552-gomcqa4wbiuielvzdtjmhpfnsc7jzfro.tugboat.qa/en/page-ii), created with German language enabled

Regarding the latter, I'm not sure about the example Node 1 link with language prefix but *without* starting slash (`de/node/1`). It puts an `en` before, result is `/en/de/node/1`, so the same issue as described by you in https://github.com/backdrop/backdrop-issues/issues/4630#issuecomment-802047167 . While it was a problem there, is this maybe expected behavior here because of the missing starting slash?

**Comment by indigoxela** [comment]
Re the `de/node/1` - yes, that's expected behavior. It's impossible to catch all and every combination. A language prefix but without leading slash is very hard to catch in a generic way. Doing so might cause more problems than it fixes.

The main thing here is to prevent false negatives in link validation - and several other issues like not relying on the link module anymore (it could be turned off) and proper handling of internal paths.

@olafgrabienski considering that, do you feel like this "works for you"?

**Comment by olafgrabienski** [comment]
> considering that, do you feel like this "works for you"?

Yes, it does!

PS, also quickly tested with success: linking a post with non ascii character in title through Editor dialog while Link module disabled

**Comment by indigoxela** [comment]
So, finally :smile: RTBC.

**Comment by quicksketch** [comment]
Thanks folks! Great work here! The implementation looks great. I think separating out the validation code to not depend on Link module was a great move. Multiple code reviews and we have tests for this PR! Very nice!

Merged https://github.com/backdrop/backdrop/pull/3552 into 1.x and 1.19.x. Thanks!
